File: 回測模組/analytics/report_generator.py
"""
報告生成模組
負責產生回測報告和分析結果
"""
import logging
from typing import List
import pandas as pd
from strategy_modules.position_manager import TradeRecord

logger = logging.getLogger(__name__)


class ReportGenerator:
    """報告生成器"""

    def __init__(self, entry_checker=None):
        """
        初始化報告生成器

        Args:
            entry_checker: 進場檢查器，用於取得進場訊號資訊
        """
        self.entry_checker = entry_checker

    def generate_report(self, df: pd.DataFrame, trades: List[TradeRecord], stock_id: str, date: str) -> None:
        """
        產生回測報告

        Args:
            df: 交易資料 DataFrame
            trades: 交易記錄列表
            stock_id: 股票代碼
            date: 交易日期
        """
        logger.info(f"\n{'='*60}")
        logger.info(f"回測報告 - {stock_id} ({date})")
        logger.info(f"{'='*60}")

        # 交易統計
        logger.info(f"\n交易統計:")
        logger.info(f"  總交易次數: {len(trades)}")

        # 詳細交易記錄
        self._log_trade_details(trades)

        # 進場訊號統計
        self._log_entry_signals()

    def _log_trade_details(self, trades: List[TradeRecord]) -> None:
        """
        輸出詳細交易記錄

        Args:
            trades: 交易記錄列表
        """
        for i, trade in enumerate(trades, 1):
            logger.info(f"\n交易 #{i}:")
            logger.info(f"  進場: {trade.entry_time} @ {trade.entry_price:.2f} (Ratio: {trade.entry_ratio:.1f})")

            if trade.partial_exit_time:
                logger.info(f"  減碼: {trade.partial_exit_time} @ {trade.partial_exit_price:.2f}")

            if trade.reentry_time:
                logger.info(f"  回補: {trade.reentry_time} @ {trade.reentry_price:.2f}")

            # 處理移動停利出場
            if hasattr(trade, 'trailing_exit_details') and trade.trailing_exit_details:
                for exit in trade.trailing_exit_details:
                    logger.info(f"  移動停利: {exit['time']} @ {exit['price']:.2f} ({exit['ratio']*100:.0f}%)")

            if trade.final_exit_time:
                logger.info(f"  清倉: {trade.final_exit_time} @ {trade.final_exit_price:.2f}")

            if trade.pnl_percent:
                logger.info(f"  收益: {trade.pnl_percent:.2f}%")

    def _log_entry_signals(self) -> None:
        """輸出進場訊號統計"""
        if self.entry_checker and hasattr(self.entry_checker, 'entry_signals'):
            if self.entry_checker.entry_signals:
                signal_df = self.entry_checker.get_entry_signals_summary()
                if not signal_df.empty:
                    logger.info("\n進場訊號分析:")
                    logger.info(f"  總進場訊號數: {len(signal_df)}")

                    # 檢查 passed 欄位存在並統計
                    if 'passed' in signal_df.columns:
                        passed_count = signal_df['passed'].sum()
                        passed_percentage = (passed_count / len(signal_df) * 100) if len(signal_df) > 0 else 0
                        logger.info(f"  通過訊號數: {passed_count}/{len(signal_df)} ({passed_percentage:.1f}%)")

                    # 顯示可用欄位（debug用）
                    logger.debug(f"  可用欄位: {list(signal_df.columns)}")

                    # 如果有條件欄位，統計各條件
                    if 'conditions' in signal_df.columns and len(signal_df) > 0:
                        # 取第一筆的條件作為範例
                        first_conditions = signal_df.iloc[0]['conditions']
                        if isinstance(first_conditions, dict):
                            logger.info("\n各條件統計:")
                            for condition_name in first_conditions.keys():
                                logger.info(f"  {condition_name}")

    def generate_summary_report(self, results: List[dict], date: str) -> pd.DataFrame:
        """
        產生批次回測總結報告

        Args:
            results: 批次回測結果列表
            date: 回測日期

        Returns:
            總結報告 DataFrame
        """
        if not results:
            logger.warning("無回測結果可供生成報告")
            return pd.DataFrame()

        # 建立 DataFrame
        df_results = pd.DataFrame(results)

        # 計算總計
        total_trades = df_results['進場次數'].sum()
        total_pnl = df_results['損益'].sum()
        avg_win_rate = df_results[df_results['進場次數'] > 0]['勝率(%)'].mean()

        logger.info(f"\n{'='*60}")
        logger.info(f"批次回測總結 - {date}")
        logger.info(f"{'='*60}")
        logger.info(f"總股票數: {len(results)}")
        logger.info(f"總交易次數: {total_trades}")
        logger.info(f"總損益: {total_pnl:,.0f}")
        logger.info(f"平均勝率: {avg_win_rate:.2f}%")

        return df_results

File: 回測模組/CSharp/BacktestModule/Analytics/ReportGenerator.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Microsoft.Extensions.Logging;
using BacktestModule.Core;

namespace BacktestModule.Analytics
{
    /// <summary>
    /// Report generator.
    /// Produces backtest reports and analysis results to the log output.
    /// </summary>
    public class ReportGenerator : IReportGenerator
    {
        private readonly ILogger<ReportGenerator> _logger;
        private readonly IEntryChecker _entryChecker;

        /// <summary>
        /// Initializes the report generator.
        /// </summary>
        /// <param name="entryChecker">Entry checker for retrieving entry signal information. Can be null.</param>
        /// <param name="logger">Logger instance. Can be null (will use NullLogger).</param>
        public ReportGenerator(IEntryChecker entryChecker = null, ILogger<ReportGenerator> logger = null)
        {
            _entryChecker = entryChecker;
            _logger = logger ?? Microsoft.Extensions.Logging.Abstractions.NullLogger<ReportGenerator>.Instance;
        }

        /// <summary>
        /// Generates a backtest report for a single stock, logging trade details and entry signal statistics.
        /// </summary>
        /// <param name="data">Tick data list.</param>
        /// <param name="trades">List of completed trade records.</param>
        /// <param name="stockId">Stock code.</param>
        /// <param name="date">Trading date string.</param>
        public void GenerateReport(List<TickData> data, List<TradeRecord> trades, string stockId, string date)
        {
            _logger.LogInformation("");
            _logger.LogInformation(new string('=', 60));
            _logger.LogInformation("Backtest Report - {StockId} ({Date})", stockId, date);
            _logger.LogInformation(new string('=', 60));

            // Trade statistics
            _logger.LogInformation("");
            _logger.LogInformation("Trade Statistics:");
            _logger.LogInformation("  Total trades: {TradeCount}", trades.Count);

            // Detailed trade records
            LogTradeDetails(trades);

            // Entry signal statistics
            LogEntrySignals();
        }

        /// <summary>
        /// Logs detailed trade records including entries, partial exits, reentries, trailing stops, and final exits.
        /// </summary>
        /// <param name="trades">List of trade records.</param>
        private void LogTradeDetails(List<TradeRecord> trades)
        {
            for (int i = 0; i < trades.Count; i++)
            {
                var trade = trades[i];
                _logger.LogInformation("");
                _logger.LogInformation("Trade #{TradeNum}:", i + 1);
                _logger.LogInformation("  Entry: {EntryTime} @ {EntryPrice:F2} (Ratio: {EntryRatio:F1})",
                    trade.EntryTime, trade.EntryPrice, trade.EntryRatio);

                if (trade.PartialExitTime.HasValue)
                {
                    _logger.LogInformation("  Partial exit: {PartialExitTime} @ {PartialExitPrice:F2}",
                        trade.PartialExitTime, trade.PartialExitPrice);
                }

                if (trade.ReentryTime.HasValue)
                {
                    _logger.LogInformation("  Reentry: {ReentryTime} @ {ReentryPrice:F2}",
                        trade.ReentryTime, trade.ReentryPrice);
                }

                // Handle trailing stop exits
                if (trade.TrailingExitDetails != null && trade.TrailingExitDetails.Count > 0)
                {
                    foreach (var exit in trade.TrailingExitDetails)
                    {
                        double price = exit.ContainsKey("price") ? Convert.ToDouble(exit["price"]) : 0.0;
                        double ratio = exit.ContainsKey("ratio") ? Convert.ToDouble(exit["ratio"]) : 0.0;
                        object time = exit.ContainsKey("time") ? exit["time"] : null;

                        _logger.LogInformation("  Trailing stop: {Time} @ {Price:F2} ({Ratio:F0}%)",
                            time, price, ratio * 100);
                    }
                }

                if (trade.FinalExitTime.HasValue)
                {
                    _logger.LogInformation("  Close all: {FinalExitTime} @ {FinalExitPrice:F2}",
                        trade.FinalExitTime, trade.FinalExitPrice);
                }

                if (trade.PnlPercent.HasValue)
                {
                    _logger.LogInformation("  PnL: {PnlPercent:F2}%", trade.PnlPercent);
                }
            }
        }

        /// <summary>
        /// Logs entry signal statistics from the entry checker.
        /// </summary>
        private void LogEntrySignals()
        {
            if (_entryChecker == null)
                return;

            var signals = _entryChecker.GetEntrySignals();
            if (signals == null || signals.Count == 0)
                return;

            _logger.LogInformation("");
            _logger.LogInformation("Entry Signal Analysis:");
            _logger.LogInformation("  Total entry signals: {SignalCount}", signals.Count);

            int passedCount = signals.Count(s => s.Passed);
            double passedPercentage = signals.Count > 0 ? (double)passedCount / signals.Count * 100 : 0;
            _logger.LogInformation("  Passed signals: {Passed}/{Total} ({Percentage:F1}%)",
                passedCount, signals.Count, passedPercentage);

            // Show condition names from the first signal
            if (signals.Count > 0 && signals[0].Conditions != null && signals[0].Conditions.Count > 0)
            {
                _logger.LogInformation("");
                _logger.LogInformation("Condition Statistics:");
                foreach (var conditionName in signals[0].Conditions.Keys)
                {
                    _logger.LogInformation("  {ConditionName}", conditionName);
                }
            }
        }

        /// <summary>
        /// Generates a batch backtest summary report.
        /// </summary>
        /// <param name="results">List of per-stock statistics dictionaries.</param>
        /// <param name="date">Backtest date string.</param>
        /// <returns>The same results list (for chaining or further processing). Returns empty list if no results.</returns>
        public List<Dictionary<string, object>> GenerateSummaryReport(
            List<Dictionary<string, object>> results, string date)
        {
            if (results == null || results.Count == 0)
            {
                _logger.LogWarning("No backtest results available for report generation.");
                return new List<Dictionary<string, object>>();
            }

            // Calculate totals
            int totalTrades = results.Sum(r =>
                r.ContainsKey("\u9032\u5834\u6B21\u6578") ? Convert.ToInt32(r["\u9032\u5834\u6B21\u6578"]) : 0);

            double totalPnl = results.Sum(r =>
                r.ContainsKey("\u640D\u76CA") ? Convert.ToDouble(r["\u640D\u76CA"]) : 0);

            var withTrades = results
                .Where(r => r.ContainsKey("\u9032\u5834\u6B21\u6578") && Convert.ToInt32(r["\u9032\u5834\u6B21\u6578"]) > 0)
                .ToList();

            double avgWinRate = withTrades.Count > 0
                ? withTrades.Average(r =>
                    r.ContainsKey("\u52DD\u7387(%)") ? Convert.ToDouble(r["\u52DD\u7387(%)"]) : 0)
                : 0;

            _logger.LogInformation("");
            _logger.LogInformation(new string('=', 60));
            _logger.LogInformation("Batch Backtest Summary - {Date}", date);
            _logger.LogInformation(new string('=', 60));
            _logger.LogInformation("Total stocks: {StockCount}", results.Count);
            _logger.LogInformation("Total trades: {TotalTrades}", totalTrades);
            _logger.LogInformation("Total PnL: {TotalPnl:N0}", totalPnl);
            _logger.LogInformation("Average win rate: {AvgWinRate:F2}%", avgWinRate);

            return results;
        }
    }
}
